File: interviews/leetcode/binary-tree-dfs/236. Lowest Common Ancestor of a Binary Tree/solution.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Function to find the lowest common ancestor
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	// Base case: if root is nil or root is either p or q
	if root == nil || root == p || root == q {
		return root
	}

	// Recursively find LCA in left and right subtrees
	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)

	// If p and q are found in different subtrees, root is the LCA
	if left != nil && right != nil {
		return root
	}

	// Otherwise, return the non-null child (either left or right)
	if left != nil {
		return left
	}
	return right
}

func main() {
	// Example test case (this will be commented out)
	// root := &TreeNode{Val: 3, Left: &TreeNode{Val: 5}, Right: &TreeNode{Val: 1}}
	// p := root.Left // 5
	// q := root.Right // 1
	// fmt.Println(lowestCommonAncestor(root, p, q)) // Output: 3
}

File: interviews/leetcode/binary-tree-dfs/236. Lowest Common Ancestor of a Binary Tree/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:

        # Base case: if root is None or root is one of the nodes
        if not root or root == p or root == q:
            return root

        # Recursively search for p and q in the left and right subtrees
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)

        # If p and q are found in different subtrees, root is their LCA
        if left and right:
            return root

        # Otherwise, return the non-null child (either left or right)
        return left if left else right
